add logging dispatcher mock

ppapapetrou76 opened the pull request:
## Description
Adds an auto-generated mockery-based mock for logging.Dispatcher interface

## Motivation and Context
We need to write better Unit tests for this interface

tobio reviewed (approved):
❤️ for `MockHelper` too :)

karencfv commented on mocks/pkg/logging/MockHelper.go:
```diff
@@ -0,0 +1,47 @@
+	}
+}
+
+// DebugLogMessageMatcher returns matcher function to be used by Mockery mock expectations of a debug message
```
```suggestion
// DebugLogMessageMatcher returns a matcher function to be used by a Mockery mock which expects a debug message.
```

karencfv commented on mocks/pkg/logging/MockHelper.go:
```diff
@@ -0,0 +1,47 @@
+	return logMessageMatcher(message, logging.DEBUG)
+}
+
+// InfoLogMessageMatcher returns matcher function to be used by Mockery mock expectations of a info message
```
```suggestion
// InfoLogMessageMatcher returns a matcher function to be used by a Mockery mock which expects an info message.
```

karencfv commented on mocks/pkg/logging/MockHelper.go:
```diff
@@ -0,0 +1,47 @@
+	return logMessageMatcher(message, logging.INFO)
+}
+
+// WarnLogMessageMatcher returns matcher function to be used by Mockery mock expectations of a warning message
```
```suggestion
// WarnLogMessageMatcher returns a matcher function to be used by a Mockery mock which expects a warning message.
```

karencfv commented on mocks/pkg/logging/MockHelper.go:
```diff
@@ -0,0 +1,47 @@
+	return logMessageMatcher(message, logging.WARNING)
+}
+
+// ErrorLogMessageMatcher returns matcher function to be used by Mockery mock expectations of an error message
```
```suggestion
// ErrorLogMessageMatcher returns a matcher function to be used by a Mockery mock which expects an error message.
```

karencfv commented on mocks/pkg/logging/Dispatcher.go:
```diff
@@ -0,0 +1,71 @@
+// specific language governing permissions and limitations
+// under the License.
+
+// Code generated by mockery v2.2.1. DO NOT EDIT.
```
Since we're introducing generated code, it's a good idea to write up a bit of documentation as to how to generate this code (what tool to use etc) and/or a make target which generates it for you if there is going to be a lot of generated code in the future.

karencfv reviewed (commented):
Thanks for the PR @ppapapetrou76 🙇‍♀️  Given that this code is specific to the logging package it's probably best to put it in there (`pkg/logging/`). 

It's also a good idea to use lowercase letters for the file names for consistency. (`Dispatcher.go` -> `dispatcher.go` and `MockHelper.go` -> `mock_helper.go`)

There is already a `mocks.go` file there. Is this new code going to replace that? if so, let's write up something about that as a disclaimer.

ppapapetrou76 commented:
Hey @karencfv I have done the following 
- Added a make target to generate mocks (also added a new dep)
- Applied your doc suggestions
- Moved the files to the same package as the actual interfaces
- Added a section in the contribution doc about mocking, how to generate and use mocks

Let me know if we are ok 🙇🏾‍♀️

karencfv commented on CONTRIBUTING.md:
````diff
@@ -243,3 +245,55 @@ $ make unit TEST_UNIT_PACKAGE=github.com/elastic/cloud-sdk-go/pkg/api
 ok      github.com/elastic/cloud-sdk-go/pkg/api 1.277s  coverage: 89.0% of statements
 ```
+
+### Generating mocks
````
I don't think we've talked about introducing these for the entire SDK 🤔 let's leave it for the logging package only for now :)
```suggestion
### Generating mocks for the logging package
```

ppapapetrou76 replied:
ok it sounds weird who want to contribute to use `mockery` only for a single package and especially a package that has already the mocks generated and it's changed once in a year. 
Any reason we don't want to use `mockery`?

Also the `generate-mocks` make target is not restricted to this package. It can generate mocks for any given package, so it's a bit confusing to have this as part of our documentaiton

karencfv replied:
I see what you mean about how it might sound strange to recommend using for a single package, I was talking specifically about this sentence:
> Any new mocks should be generated with Mockery

While I don't want to block tooling from adding mocks in whichever way you find most suitable to test the logging package, which is mostly maintained/used by tooling. I also do not think that we should commit to only using Mockery for all of our mocks. My reasons are as follows:
- For our apivalidation tests we are already using [prism](https://github.com/stoplightio/prism).
- As far as I know no one in delivery has dug too deeply in Mockery to confirm at this moment that we are OK committing to using Mockery exclusively.

If you change the above sentence to say something along the lines of `An option to generate mocks is to use Mockery via the make generate-mocks target...` I would be OK with that 🙂  .

karencfv commented on CONTRIBUTING.md:
```diff
@@ -243,3 +245,55 @@ $ make unit TEST_UNIT_PACKAGE=github.com/elastic/cloud-sdk-go/pkg/api
+    m.On("SimpleReceiver").Return(true)
+
+    // Setup a return value for specific params
+    m.On("ComplexReceiver", "I wanna know what love is", Params{ WhoShouldShowMe: "you" }).Return(true, nil)
```
😄

karencfv commented on CONTRIBUTING.md:
```diff
@@ -243,3 +245,55 @@ $ make unit TEST_UNIT_PACKAGE=github.com/elastic/cloud-sdk-go/pkg/api
+
+### Generating mocks
+
+*Any new mocks should be generated with [Mockery](https://github.com/vektra/mockery) via `make generate-mocks`. New code should utilise these mocks.*
```
```suggestion
*Any new mocks for the logging package should be generated with [Mockery](https://github.com/vektra/mockery) via `make generate-mocks`. New code should utilise these mocks.*
```

karencfv reviewed (commented):
Thanks a bunch for addressing my comments and the documentation! 🙇‍♀️ Just left a couple minor things.

Btw, is this file still relevant https://github.com/elastic/cloud-sdk-go/blob/master/pkg/logging/mocks.go ? Is this going to be replaced with new tests?

ppapapetrou76 commented:
> Btw, is this file still relevant https://github.com/elastic/cloud-sdk-go/blob/master/pkg/logging/mocks.go ? Is this going to be replaced with new tests?

Yeah eventually. We have a shit tone of tests using the old mocks so we can't remove it just like this. 😞

karencfv commented:
> Yeah eventually. We have a shit tone of tests using the old mocks so we can't remove it just like this. 😞

Nw! This was more of a question than an ask :) I realise this is the first step of more work to come.

ppapapetrou76 commented:
Closing in favor of https://github.com/elastic/internal-cloud-sdk-go/pull/259
